database/raft/fsync_darwin.go

```diff
@@ -0,0 +1,16 @@
+// +build darwin
+
+package raft
+
+import (
+	"os"
+	"syscall"
+)
+
+func fsync(f *os.File) error {
+	_, _, errno := syscall.Syscall(syscall.SYS_FCNTL, f.Fd(), syscall.F_FULLFSYNC, 0)
```

Comment:
I'll defer to the explanation in [Apple's fsync man page](https://developer.apple.com/legacy/library/documentation/Darwin/Reference/ManPages/man2/fsync.2.html):

     Note that while fsync() will flush all data from the host to the drive (i.e. the "permanent storage
     device"), the drive itself may not physically write the data to the platters for quite some time and it
     may be written in an out-of-order sequence.

     Specifically, if the drive loses power or the OS crashes, the application may find that only some or
     none of their data was written.  The disk drive may also re-order the data so that later writes may be
     present, while earlier writes are not.

     This is not a theoretical edge case.  This scenario is easily reproduced with real world workloads and
     drive power failures.

     For applications that require tighter guarantees about the integrity of their data, Mac OS X provides
     the F_FULLFSYNC fcntl.  The F_FULLFSYNC fcntl asks the drive to flush all buffered data to permanent
     storage.  Applications, such as databases, that require a strict ordering of writes should use F_FULLF-
     SYNC to ensure that their data is written in the order they expect.  Please see fcntl(2) for more
     detail.

(Personally I think it would be better if the default were the safe way perhaps with a flag to turn it off, and even then the flag should arguably be a system-level option rather than an API call parameter. But it is what it is.)

Linux [until about 2008 it seems](https://www.mail-archive.com/linux-fsdevel@vger.kernel.org/msg12025.html) didn't provide the safer option at all, and now I think it only provides the safer option? It's hard to get a straight answer out of the Internet.

Basically I tried to use the safest available option that I'm aware of on each platform. Incidentally, etcd itself does the same thing https://github.com/coreos/etcd/blob/2724c3946eb2f3def5ed38a127be982b62c81779/pkg/fileutil/sync_darwin.go.